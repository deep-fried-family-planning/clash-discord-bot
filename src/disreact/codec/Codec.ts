import {Intrinsic} from '#src/disreact/codec/elem/index.ts';
import {Keys} from '#src/disreact/codec/elem/keys.ts';
import {DisReactConfig} from '#src/disreact/utils/DisReactConfig.ts';
import {E, S} from '#src/disreact/utils/re-exports.ts';
import {Rx} from '#src/disreact/codec/rx.ts';
import {Tx} from '#src/disreact/codec/tx.ts';

export class Codec extends E.Service<Codec>()('disreact/Codec', {
  effect: DisReactConfig.use(() => {
    return {
      primitive     : Keys.primitive,
      normalization : Intrinsic.NORM,
      encoding      : Intrinsic.ENC,
      decodeRequest : S.decodeSync(Rx.Request),
      encodeResponse: S.encodeSync(Tx.Response),
    };
  }),
  accessors: true,
}) {}
