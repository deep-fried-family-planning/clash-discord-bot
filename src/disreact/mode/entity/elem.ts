import type * as Event from '#src/disreact/mode/entity/event.ts';
import * as FC from '#src/disreact/mode/entity/fc.ts';
import * as Props from '#src/disreact/mode/entity/props.ts';
import * as Array from 'effect/Array';
import * as Data from 'effect/Data';

export declare namespace Elem {
  export type Val = {
    type : false;
    props: any;
    nodes: Any[];
  };
  export type Api = {
    id?     : string;
    type    : string;
    props   : Props.Props;
    nodes   : Any[];
    handler?: Event.Handler | undefined;
  };
  export type Fn = {
    id?  : string;
    type : FC.FC.Any<any, any>;
    props: Props.Props;
    nodes: Any[];
  };
  export type Node = Api | Fn;
  export type Any = Val | Node;
  export type Children = Any | Children[];
  export type Elem = Any;
}
export type Elem = Elem.Any;
export type Val = Elem.Val;
export type Api = Elem.Api;
export type Fn = Elem.Fn;
export type Node = Elem.Node;
export type Any = Elem.Any;
export type Children = Elem.Children;

const emptyNodes = () => Data.array([] as Elem.Any[]) as Elem.Any[];

export const ensureChildren = (cs?: Elem.Any | Elem.Any[]): Elem.Any[] => {
  if (!cs) return emptyNodes();
  if (Array.isArray(cs)) return Data.array(cs.flat()) as Elem.Any[];
  return Data.array([cs]) as Elem.Any[];
};

export const isVal = (e: Elem.Any): e is Elem.Val => !e.type;
export const isApi = (e: Elem.Any): e is Elem.Api => typeof e.type === 'string';
export const isFn = (e: Elem.Any): e is Elem.Fn => FC.isFC(e.type);

export const val = (value: any): Elem.Val => {
  return Data.struct({
    type : false,
    props: value,
    nodes: emptyNodes(),
  });
};

export const api = (tag: string, props: Props.Props): Elem.Api => {
  const handler = Props.extractHandler(props);
  return Data.struct({
    type : tag,
    props: Props.make(props),
    nodes: ensureChildren(props.children),
    handler,
  });
};

export const fn = (tag: any, props: Props.Props): Elem.Fn => {
  return Data.struct({
    type : FC.make(tag),
    props: Props.make(props),
    nodes: emptyNodes(),
  });
};
