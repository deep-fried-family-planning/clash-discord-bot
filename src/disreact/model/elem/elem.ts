import {Props} from '#src/disreact/model/elem/props.ts';
import type {Trigger} from '#src/disreact/model/elem/trigger.ts';
import {FC} from '#src/disreact/model/elem/fc.ts';
import {Fibril} from '#src/disreact/model/meta/fibril.ts';
import {E} from '#src/disreact/utils/re-exports.ts';

export * as Elem from '#src/disreact/model/elem/elem.ts';
export type Elem =
  | Task
  | Rest
  | Fragment
  | Value;

export interface Node {
  type   : any;
  id?    : string | undefined;
  ids?   : string | undefined;
  idn?   : string | undefined;
  idx?   : number | undefined;
  length?: number | undefined;
  fibril?: Fibril | undefined;
  props  : Props;
  nodes  : Elem[];
}

export const isNode = (self: any): self is Node => typeof self === 'object';

export type Child =
  | Task
  | Rest
  | Fragment
  | ValueType;

export type Children = Child | Child[];

export const isChild = (children: any): children is Child => !children.length;

export const connectChildren = (parent: Node, children: Children): Elem[] => {
  if (isValueType(children)) return [makeValue(children)];
  if (isChild(children)) return [connectChild(parent, children)];
  return connectNodes(parent, children);
};

export const connectNodes = (parent: Node, children: Child[]): Elem[] => {
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    children[i] = isValueType(child)
      ? makeValue(child)
      : connectChild(parent, child, i);
  }
  return children as Elem[];
};

export const connectChild = (parent: Node, child: Node, idx: number = 0): Elem => {
  child.idx = idx;
  const child_id = `${child.idn ?? child.type ?? ''}:${idx}`;
  const parent_id = `${parent.idn ?? parent.type ?? ''}:${parent.idx}`;
  child.id = `${parent.id}:${child_id}`;
  child.ids = `${parent_id}:${child_id}`;
  return child;
};

export type TaskType<E = any, R = any> = (prop?: any) =>
  | Child
  | Promise<Child>
  | E.Effect<Child, E, R>;

export interface Task extends Node {
  type  : FC;
  fibril: Fibril;
}

export const isTaskType = (self: any): self is TaskType => typeof self === 'function';

export const isTask = (self: any): self is Task => typeof self === 'object' && typeof self.type === 'function';

export const isEqualTask = (a: Task, b: Task) => {
  if (a === b) return true;
  if (a.type !== b.type) return false;
  if (!Props.isEqual(a.props, b.props)) return false;
  if (a.nodes.length !== b.nodes.length) return false;
  return true;
};

export const makeTask = (type: any, props: any): Task => {
  const fc = FC.make(type);
  return {
    type  : fc,
    idn   : FC.getName(fc),
    props,
    nodes : [],
    fibril: Fibril.make(),
  };
};

export const cloneTask = (self: Task): Task => {
  const {type, props, nodes, fibril, ...rest} = self;
  const cloned = structuredClone(rest) as Task;
  cloned.type = type;
  cloned.fibril = Fibril.clone(fibril);
  cloned.props = Props.deepCloneTaskProps(props);
  cloned.nodes = nodes;
  return cloned;
};

export type RestType = string;

export interface Rest extends Node {
  type    : string;
  handler?: Trigger.Handler<any> | undefined;
}

export const isRestType = (self: any): self is RestType => typeof self === 'string';

export const isRest = (self: any): self is Rest => typeof self === 'object' && typeof self.type === 'string';

export const makeRest = (type: string, props: any, nodes: any[]): Rest =>
  ({
    type,
    props,
    nodes,
    handler: Props.getHandler(props),
  });

export const cloneRest = (self: Rest): Rest => {
  const {props, nodes, handler, ...rest} = self;
  const cloned = structuredClone(rest) as Rest;
  cloned.props = Props.cloneKnownProps(props);
  cloned.nodes = nodes;
  cloned.handler = handler;
  return cloned;
};

/**
 * Fragment
 */
export type FragmentType = undefined;

export const FragmentType = undefined;

export interface Fragment extends Node {
  type: FragmentType;
};

export const isFragmentType = (self: any): self is FragmentType => self === FragmentType;

export const isFragment = (self: any): self is Fragment => self.type === FragmentType;

export const makeFragment = (type: undefined, props: any): Fragment =>
  ({
    type,
    props,
    nodes: props.children,
  });

export const cloneFragment = (self: Fragment) => {
  const {props, nodes, ...rest} = self;
  const cloned = structuredClone(rest) as Fragment;
  cloned.props = Props.cloneKnownProps(props);
  cloned.nodes = nodes;
  return cloned;
};

/**
 * Value
 */
export type ValueType =
  | string
  | bigint
  | number
  | boolean
  | null
  | undefined
  | symbol;

export const isValueType = (type: any): type is ValueType =>
  !type ||
  typeof type !== 'object';

export type Value =
  | string
  | bigint
  | number
  | true;

export const isValue = (self: Elem): self is Value =>
  typeof self !== 'object' ||
  self === null;

export const makeValue = (type: any): Value => {
  return type;
};

export const cloneValue = (self: Value) => structuredClone(self) as Value;
