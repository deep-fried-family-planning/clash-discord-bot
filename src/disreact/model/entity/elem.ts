import {Keys} from '#src/disreact/codec/rest-elem/keys';
import {Fibril} from '#src/disreact/model/entity/fibril.ts';
import type {Trigger} from '#src/disreact/model/entity/trigger.ts';
import {Data, Differ} from 'effect';
import {FC} from 'src/disreact/model/entity/fc.ts';

const HANDLER_KEYS = [
  Keys.onclick,
  Keys.onselect,
  Keys.onsubmit,
];

const RESERVED = [
  ...HANDLER_KEYS,
  Keys.children,
  Keys.handler,
];

export const TypeId = Symbol('disreact/Elem');
export type TypeId = typeof TypeId;

export type Return = Elem | Elem[];

export interface MetaProps {
  [TypeId]: TypeId;
  type    : any;
  id?     : string | undefined;
  ids?    : string | undefined;
  idn?    : string | undefined;
  idx?    : number | undefined;
  props   : any;
  nodes   : Any[];
}

/**
 * Primitive
 */
export type Value =
  | string
  | bigint
  | number
  | boolean
  | null
  | undefined
  | symbol;

export const isValue = (self: unknown): self is Value => {
  if (!self) return true;
  if (typeof self === 'object') return false;
  if (typeof self === 'function') return false;
  return true;
};

export const makeValue = (value: any): Value => value;

export const cloneValue = (self: Value): Value => structuredClone(self);

/**
 * Fragment
 */
export type Frag = undefined;

export const Frag = undefined;

export const makeFragment = (type: undefined, props: any) => {
  return props.children;
};

/**
 * Rest
 */
export interface Rest extends MetaProps {
  type   : string;
  handler: Trigger.Handler<any>;
}

export const isRest = (self: any): self is Rest =>
  typeof self === 'object' &&
  typeof self.type === 'string';

export const makeRest = (type: string, props: any, nodes: any[]): Rest => {
  const rest = {
    type,
    props,
    nodes,
  } as Rest;

  for (let i = 0; i < HANDLER_KEYS.length; i++) {
    const key = HANDLER_KEYS[i];
    const handler = props[key];

    if (handler) {
      rest.handler = handler;
      delete props[key];
    }
  }

  return rest;
};

export const jsxRest = (type: string, props: any): Rest => {
  const child = props.children;
  delete props.children;
  return makeRest(type, props, child ? [child] : []);
};

export const jsxsRest = (type: string, props: any): Rest => {
  const nodes = props.children;
  delete props.children;
  return makeRest(type, props, nodes);
};

export const cloneRest = (self: Rest): Rest => {
  const {props, nodes, handler, ...rest} = self;

  const reserved = {} as any;

  for (const key of RESERVED) {
    const prop = props[key];
    if (prop) {
      reserved[key] = prop;
      delete props[key];
    }
  }

  const cloned = structuredClone(rest) as Rest;
  cloned.props = structuredClone(props);
  cloned.nodes = nodes;
  cloned.handler = handler;

  for (const key of Object.keys(reserved)) {
    cloned.props[key] = reserved[key];
    props[key] = reserved[key];
  }

  return cloned;
};

/**
 * Task
 */
export interface Task extends MetaProps {
  type  : FC;
  fibril: Fibril;
}

export const jsxTask = (type: any, props: any): Task => {
  const fc = FC.make(type);
  const task = {} as Task;
  task.idn = FC.getName(fc);
  task.type = fc;
  task.props = props;
  task.nodes = [] as Elem[];
  task.fibril = Fibril.make();
  return task;
};

export const jsxsTask = (type: any, props: any): Task => {
  const init = jsxTask(type, props);
  return init;
};

export const isTask = (self: unknown): self is Task => {
  switch (typeof self) {
    case 'object':
      return typeof (self as any).type === 'function';
  }
  return false;
};

export const cloneTask = (self: Task): Task => {
  const {props, fibril, type, nodes, id, idx} = self;
  const clonedProps = deepCloneTaskProps(props);
  const task = jsxTask(type, clonedProps);
  task.fibril = Fibril.cloneStrand(fibril);
  task.nodes = nodes;
  task.id = id;
  task.idx = idx;
  return task;
};

const deepCloneTaskProps = (props: any): any => {
  if (!props) {
    return props;
  }

  try {
    return structuredClone(props);
  }
  catch (e) {/**/}

  const acc = {} as any;

  for (const key of Object.keys(props)) {
    const original = props[key];
    const originalType = typeof original;

    if (originalType === 'object') {
      if (!original) {
        acc[key] = null;
        continue;
      }
      if (Array.isArray(original)) {
        acc[key] = original.map((item) => deepCloneTaskProps(item));
        continue;
      }
      acc[key] = deepCloneTaskProps(original);
      continue;
    }

    acc[key] = original;
  }

  return props;
};

/**
 * Elem
 */
export * as Elem from '#src/disreact/model/entity/elem.ts';
export type Elem =
  | Rest
  | Task;

export type Any =
  | Elem
  | Value;

export const jsx = (type: any, props: any): Elem => {
  if (type === undefined) {
    return props.children;
  }

  switch (typeof type) {
    case 'string':
      return jsxRest(type, props);
    case 'function':
      return jsxTask(type, props);
    default:
      throw new Error();
  }
};

export const jsxs = (type: any, props: any): Elem => {
  props.children = props.children.flat();

  if (type === undefined) {
    return props.children;
  }

  switch (typeof type) {
    case 'string':
      return jsxsRest(type, props);
    case 'function':
      return jsxsTask(type, props);
    default:
      throw new Error();
  }
};

export const jsxDEV = (type: any, props: any): Elem => {
  if (!Array.isArray(props.children)) {
    return jsx(type, props);
  }
  return jsxs(type, props);
};

export const cloneElem = (self: Elem) => {
  if (isRest(self)) {
    return cloneRest(self);
  }

  return cloneTask(self);
};

export const deepCloneElem = <A extends Any>(elem: A): A => {
  if (isValue(elem)) {
    return structuredClone(elem) as A;
  }

  const cloned = cloneElem(elem);

  for (let i = 0; i < cloned.nodes.length; i++) {
    const node = cloned.nodes[i];
    cloned.nodes[i] = deepCloneElem(node);
  }

  return cloned as A;
};

export const connectChild = (parent: Elem, child: Elem, idx: number) => {
  child.idx = idx;

  const child_id = `${child.idn ?? child.type}:${idx}`;
  const parent_id = `${parent.idn ?? parent.type}:${parent.idx}`;

  child.id = `${parent.id}:${child_id}`;
  child.ids = `${parent_id}:${child_id}`;
};

export type Ops = Data.TaggedEnum<{
  Skip   : {};
  Add    : {node: Any};
  Delete : {node: Any; idx: number};
  Replace: {node: Any; idx: number; child: Any};
  Update : {node: Any; idx: number; child: Any};
  Render : {};
}>;

export const Ops = Data.taggedEnum<Ops>();
const empty = Ops.Skip() as Ops;
const combine = () => {throw new Error();};

export const isSame = (a: Elem, b: Elem) => {
  if (a === b) return true;
  if (isValue(a) && isValue(b)) return false;
  if (isRest(a) && isRest(b)) {
    if (a.type !== b.type) return false;
    return true;
  }
  if (isTask(a) && isTask(b)) {
    if (a.type !== b.type) return false;
    return true;
  }
  return false;
};
