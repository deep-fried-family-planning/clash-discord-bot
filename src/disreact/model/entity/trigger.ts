import {D, E, pipe, S} from '#src/disreact/utils/re-exports.ts';
import type {Cause} from 'effect';
import {Predicate} from 'effect';

export const TypeId = Symbol('disreact/Trigger');

export class TriggerDefect extends D.TaggedError('disreact/TriggerDefect')<{
  message?: string;
  cause?  : unknown;
}> {}

export * as Trigger from '#src/disreact/model/entity/trigger.ts';
export type Trigger<A = any> = {
  id      : string;
  data    : A;
  [TypeId]: typeof TypeId;
};

export const make = (id: string, data: any): Trigger =>
  ({
    [TypeId]: TypeId,
    id,
    data,
  });

export const declare = <A, I, R>(data: S.Schema<A, I, R>) =>
  S.Struct({
    id  : S.String,
    data: data,
  });

export type Event<A> = Trigger<A>['data'];

export const HandlerTypeId = Symbol('disreact/Trigger/Handler');

export type Handler<A> = <E, R>(event: Event<A>) =>
  | void
  | Promise<void>
  | E.Effect<void, E, R>;

const isHandler = <A>(h: unknown): h is Handler<A> =>
  typeof h === 'function';

export const declareHandler = <A, I, R>(data: S.Schema<A, I, R>) =>
  S.declare(
    isHandler<typeof data.Type>,
  );

export const apply = <A = any>(handler: Handler<A>, event: Trigger<A>) =>
  pipe(
    E.try({
      try  : () => handler(event.data),
      catch: (e) => new TriggerDefect({
        message: 'Handler threw an exception',
        cause  : e,
      }),
    }),
    E.flatMap((output) => {
      if (!output) {
        return E.void;
      }

      if (E.isEffect(output)) {
        return output as E.Effect<void>;
      }

      if (Predicate.isPromise(output)) {
        return E.tryPromise({
          try  : () => output,
          catch: (e) => new TriggerDefect({
            message: 'Handler threw an exception',
            cause  : e,
          }),
        });
      }

      return E.fail(
        new TriggerDefect({
          message: 'Invalid handler output',
        }),
      ) as E.Effect<void, Cause.UnknownException | TriggerDefect>;
    }),
  );
