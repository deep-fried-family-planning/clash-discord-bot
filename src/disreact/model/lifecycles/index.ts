export * from './render.ts';
export * from './utils.ts';
export * from './invoke.ts';
import * as Element from '../../codec/element/index.ts';



export const cloneTree = Element.cloneTree;
