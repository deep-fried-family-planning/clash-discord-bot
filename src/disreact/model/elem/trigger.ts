import {Elem} from '#src/disreact/model/elem/elem.ts';
import {D, E, pipe, S} from '#src/disreact/utils/re-exports.ts';
import {Predicate} from 'effect';

export * as Trigger from '#src/disreact/model/elem/trigger.ts';
export type Trigger<A = any> = {
  id  : string;
  data: A;
};

export const make = (id: string, data: any): Trigger => ({
  id,
  data,
});

export const declare = <A, I, R>(data: S.Schema<A, I, R>) =>
  S.Struct({
    id  : S.String,
    data: data,
  });

export type Event<A> = Trigger<A>['data'];

export type Handler<A, E = any, R = any> = (event: Event<A>) =>
  | void
  | Promise<void>
  | E.Effect<void, E, R>;

const isHandler = <A>(h: unknown): h is Handler<A> =>
  typeof h === 'function';

export const declareHandler = <A, I, R>(data: S.Schema<A, I, R>) =>
  S.declare(
    isHandler<typeof data.Type>,
  );

export class TriggerDefect extends D.TaggedError('disreact/TriggerDefect')<{
  cause: unknown;
}> {}

export const isTarget = (event: Trigger, elem: Elem): elem is Elem.Rest => {
  if (Elem.isRest(elem)) {
    if (elem.props.custom_id === event.id || elem.ids === event.id) {
      return true;
    }
  }
  return false;
};

export const apply = <A = any>(handler: Handler<A>, event: Trigger<A>) =>
  pipe(
    E.sync(() => handler(event.data)),
    E.flatMap((output) => {
      if (Predicate.isPromise(output)) {
        return E.promise(async () => await output);
      }

      if (E.isEffect(output)) {
        return output as E.Effect<void>;
      }

      return E.void;
    }),
    E.catchAllDefect((e) => E.fail(e as Error)),
  );
