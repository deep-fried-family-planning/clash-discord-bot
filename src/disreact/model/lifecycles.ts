import {EH} from '#src/disreact/model/entity/eh.ts'
import {Elem} from '#src/disreact/model/entity/elem.ts'
import {Props} from '#src/disreact/model/entity/props.ts'
import {Root} from '#src/disreact/model/entity/root.ts'
import {Fibril} from '#src/disreact/model/fibril/fibril.ts'
import {Relay, relayPartial, RelayStatus} from '#src/disreact/model/Relay.ts'
import {E, ML, pipe} from '#src/disreact/re-exports.ts'
import {Lifecycle} from 'src/disreact/model/lifecycle.ts'

export * as Lifecycles from '#src/disreact/model/lifecycles.ts'
export type Lifecycles = never

export const handleEvent = (root: Root, event: any) =>
  E.suspend(() => {
    const stack = ML.make<Elem>(root.elem)

    while (ML.tail(stack)) {
      const elem = ML.pop(stack)!

      if (Elem.isRest(elem)) {
        if (elem.props.custom_id === event.id || elem.ids === event.id) {
          return pipe(
            EH.apply(elem.handler, event),
            E.andThen(() => Relay.sendStatus(RelayStatus.Handled())),
            E.tap(() => Lifecycle.notifyOnHandlePiped(root)),
          )
        }
      }

      for (let i = 0; i < elem.nodes.length; i++) {
        const node = elem.nodes[i]
        if (!Elem.isPrim(node)) {
          ML.append(stack, elem.nodes[i])
        }
      }
    }

    return E.fail(new Error('Event not handled'))
  })


export const initialize = (root: Root) =>
  pipe(
    initializeSubtree(root, root.elem),
    E.as(root),
  )

export const initializeSubtree = (root: Root, elem: Elem) => E.gen(function* () {
  const stack = ML.make<Elem>(elem)

  let hasSentPartial = false

  while (ML.tail(stack)) {
    const elem = ML.pop(stack)!

    if (Elem.isTask(elem)) {
      elem.nodes = yield* Lifecycle.renderElemPiped(root, elem)
    }
    else if (!hasSentPartial) {
      hasSentPartial = yield* relayPartial(elem)
    }

    for (let i = 0; i < elem.nodes.length; i++) {
      const node = elem.nodes[i]

      if (!Elem.isPrim(node)) {
        Elem.connectChild(elem, node, i)
        Root.mountElem(root, node)
        ML.append(stack, node)
      }
    }
  }

  return elem
})


export const hydrate = (root: Root) => pipe(
  E.iterate(ML.make<Elem>(root.elem), {
    while: (stack) => !!ML.tail(stack),
    body : (stack) =>
      pipe(
        E.suspend(() => {
          const elem = ML.pop(stack)!

          if (Elem.isTask(elem)) {
            if (root.nexus.strands[elem.id!]) {
              elem.strand = root.nexus.strands[elem.id!]
              elem.strand.nexus = root.nexus
            }

            return pipe(
              Lifecycle.renderElemPiped(root, elem),
              E.map((children) => {
                elem.nodes = children
                return elem
              }),
            )
          }

          return E.succeed(elem as Elem)
        }),
        E.map((elem) => {
          for (let i = 0; i < elem.nodes.length; i++) {
            const node = elem.nodes[i]

            if (!Elem.isPrim(node)) {
              Elem.connectChild(elem, node, i)
              Root.mountElem(root, node)
              ML.append(stack, node)
            }
          }
          return stack
        }),
      ),
  }),
  E.as(root),
)


export const rerender = (root: Root) => E.gen(function* () {
  const stack = ML.empty<[Elem, Elem.Any[]]>()
  let hasSentPartial = false

  if (Elem.isRest(root.elem)) {
    for (let i = 0; i < root.elem.nodes.length; i++) {
      const node = root.elem.nodes[i]

      if (Elem.isPrim(node)) {
        continue
      }
      else if (Elem.isTask(node)) {
        ML.append(stack, [node, yield* Lifecycle.renderElemPiped(root, node)])
      }
      else {
        ML.append(stack, [node, node.nodes])
      }
    }
  }
  else {
    ML.append(stack, [root.elem, yield* Lifecycle.renderElemPiped(root, root.elem)])
  }

  while (ML.tail(stack)) {
    const [parent, rs] = ML.pop(stack)!
    const maxlen = Math.max(parent.nodes.length, rs.length)

    for (let i = 0; i < maxlen; i++) {
      const curr = parent.nodes[i]
      const rend = rs[i]

      if (!curr) {
        if (Elem.isPrim(rend)) {
          parent.nodes[i] = rend
        }
        else {
          yield* mountSubtree(root, rend)
          parent.nodes[i] = rend
        }
      }
      else if (!rend) {
        if (Elem.isPrim(curr)) {
          delete parent.nodes[i]
        }
        else {
          dismountSubtree(root, curr)
          delete parent.nodes[i]
        }
      }

      else if (Elem.isPrim(curr)) {
        if (Elem.isPrim(rend)) {
          if (curr !== rend) {
            parent.nodes[i] = rend
          }
        }
        else if (Elem.isRest(rend)) {
          yield* mountSubtree(root, rend)
          parent.nodes[i] = rend
        }
        else {
          yield* mountSubtree(root, rend)
          parent.nodes[i] = rend
        }
      }

      else if (Elem.isRest(curr)) {
        if (!hasSentPartial) {
          hasSentPartial = yield* relayPartial(curr)
        }

        if (Elem.isPrim(rend)) {
          dismountSubtree(root, curr)
          parent.nodes[i] = rend
        }
        else if (Elem.isRest(rend)) {
          if (curr.type !== rend.type) {
            dismountSubtree(root, curr)
            yield* mountSubtree(root, rend)
            parent.nodes[i] = rend
          }
          if (!Props.isEqual(curr.props, rend.props)) {
            curr.props = rend.props
          }
          ML.append(stack, [curr, rend.nodes])
        }
        else {
          dismountSubtree(root, curr)
          yield* mountSubtree(root, rend)
          parent.nodes[i] = rend
        }
      }

      else {
        if (Elem.isPrim(rend)) { // Task => Primitive
          dismountSubtree(root, curr)
          parent.nodes[i] = rend
        }
        else if (Elem.isRest(rend) || curr.idn !== rend.idn) { // Task => Rest or Task => Task
          dismountSubtree(root, curr)
          yield* mountSubtree(root, rend)
          parent.nodes[i] = rend
        }
        else if (
          Props.isEqual(curr.props, rend.props) && // Task Changed
          Fibril.isSameStrand(curr.strand)
        ) {
          ML.append(stack, [curr, rend.nodes])
        }
        else {
          const rerendered = yield* Lifecycle.renderElemPiped(root, curr)
          ML.append(stack, [curr, rerendered])
        }
      }
    }
  }

  return root
})


export const mountSubtree = (root: Root, elem: Elem) => E.gen(function* () {
  const stack = ML.make<Elem>(elem)

  while (ML.tail(stack)) {
    const elem = ML.pop(stack)!

    if (Elem.isTask(elem)) {
      Root.mountTask(root, elem)
      elem.nodes = yield* Lifecycle.renderElemPiped(root, elem)
    }

    for (let i = 0; i < elem.nodes.length; i++) {
      const node = elem.nodes[i]

      if (!Elem.isPrim(node)) {
        ML.append(stack, node)
      }
    }
  }
})

export const dismountSubtree = (root: Root, elem: Elem) => {
  const stack = ML.make<Elem>(elem)

  while (ML.tail(stack)) {
    const elem = ML.pop(stack)!

    if (Elem.isTask(elem)) {
      Root.dismountTask(root, elem)
    }

    for (let i = 0; i < elem.nodes.length; i++) {
      const node = elem.nodes[i]

      if (!Elem.isPrim(node)) {
        ML.append(stack, node)
      }
    }
  }
}
