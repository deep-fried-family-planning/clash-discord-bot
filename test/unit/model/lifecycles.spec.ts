import {encodeRoot} from '#src/disreact/codec/Codec.ts';
import { Elem } from '#src/disreact/model/entity/elem';
import {Trigger} from '#src/disreact/model/entity/trigger';
import {Lifecycles} from '#src/disreact/model/lifecycles.ts';
import {Registry} from '#src/disreact/model/Registry.ts';
import {E} from '#src/internal/pure/effect.ts';
import {describe, expect} from '@effect/vitest';
import {pipe, Record} from 'effect';
import {TestMessage} from 'test/unit/components/test-message.tsx';
import {expectJSON, it} from 'test/unit/components/TestRegistry.tsx';



it.effect('when dispatching an event', E.fn(function* () {
  const registry = yield* Registry;
  const root = yield* registry.checkout(TestMessage);

  yield* Lifecycles.initialize(root);

  const event = Trigger.make('actions:2:button:0', {});

  expect(Record.map(root.fibrils, (v) => v.stack)).toMatchInlineSnapshot(`
    {
      "TestMessage": [
        {
          "s": 0,
        },
      ],
      "TestMessage:message:0:Header:0": [],
    }
  `);

  expect(JSON.stringify(encodeRoot(root), null, 2)).toMatchSnapshot();

  yield* Lifecycles.handleEvent(root, event);
  yield* Lifecycles.rerender(root);

  expect(Record.map(root.fibrils, (v) => v.stack)).toMatchInlineSnapshot(`
    {
      "TestMessage": [
        {
          "s": 1,
        },
      ],
      "TestMessage:message:0:Header:0": [],
    }
  `);

  expect(JSON.stringify(encodeRoot(root), null, 2)).toMatchSnapshot();
}));

describe('given event.id does not match any node.id', () => {
  it.effect('when dispatching an event', E.fn(function* () {
    const registry = yield* Registry;
    const root = yield* registry.checkout(TestMessage);
    yield* Lifecycles.initialize(root);
    yield* Lifecycles.rerender(root);

    const event = Trigger.make('buttons:1:button:0', {});

    yield* pipe(
      Lifecycles.handleEvent(root, event),
      E.catchAll((err) => {
        expect(err).toMatchInlineSnapshot(`[Error: Event not handled]`);
        return E.void;
      }),
    );
  }));
});

it.effect('when rendering an initial tree', E.fn(function* () {
  const registry = yield* Registry;
  const root = yield* registry.checkout(TestMessage);
  yield* Lifecycles.initialize(root);
  expect(JSON.stringify(encodeRoot(root), null, 2)).toMatchSnapshot();
}));

it.effect(`when hydrating an empty root (performance)`, E.fn(function* () {
  const runs = Array.from({length: 10000});
  const registry = yield* Registry;

  for (let i = 0; i < runs.length; i++) {
    const root = yield* registry.checkout(TestMessage);

    yield* Lifecycles.initialize(root);

    yield* Lifecycles.hydrate(root);

    const event = Trigger.make('actions:2:button:0', {});

    yield* Lifecycles.handleEvent(root, event);
    yield* Lifecycles.rerender(root);

    const encoded = encodeRoot(root);
  }
}), {timeout: 10000});
