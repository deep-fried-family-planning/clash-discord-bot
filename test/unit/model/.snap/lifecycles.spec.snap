// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when dispatching an event 2`] = `
"{
  "embeds": [
    {
      "title": "Omni Board",
      "description": "V2 - JSX Pragma"
    },
    {
      "title": "Test Title - NOT markdown",
      "description": "### H3-# \\n@123456\\n__$ not a link__"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "label": "Start 0",
          "style": 1,
          "type": 2,
          "custom_id": "actions:2:button:0"
        },
        {
          "label": "Help",
          "style": 2,
          "type": 2,
          "custom_id": "actions:2:secondary:1"
        }
      ]
    },
    {
      "type": 1,
      "components": [
        {
          "custom_id": "message:0:select:3",
          "type": 3,
          "options": [
            {
              "value": "1",
              "label": "1"
            },
            {
              "value": "2",
              "label": "2"
            },
            {
              "value": "3",
              "label": "3",
              "default": true
            }
          ]
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when dispatching an event 4`] = `
"{
  "embeds": [
    {
      "title": "Test Title - NOT markdown",
      "description": "### H3-# \\n@123456\\n__$ not a link__"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "label": "Start 1",
          "style": 1,
          "type": 2,
          "custom_id": "actions:2:button:0"
        },
        {
          "label": "Help",
          "style": 2,
          "type": 2,
          "custom_id": "actions:2:secondary:1"
        }
      ]
    },
    {
      "type": 1,
      "components": [
        {
          "custom_id": "message:0:select:3",
          "type": 3,
          "options": [
            {
              "value": "1",
              "label": "1"
            },
            {
              "value": "2",
              "label": "2"
            },
            {
              "value": "3",
              "label": "3",
              "default": true
            }
          ]
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when rendering an initial tree 1`] = `
"{
  "embeds": [
    {
      "title": "Omni Board",
      "description": "V2 - JSX Pragma"
    },
    {
      "title": "Test Title - NOT markdown",
      "description": "### H3-# \\n@123456\\n__$ not a link__"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "label": "Start 0",
          "style": 1,
          "type": 2,
          "custom_id": "actions:2:button:0"
        },
        {
          "label": "Help",
          "style": 2,
          "type": 2,
          "custom_id": "actions:2:secondary:1"
        }
      ]
    },
    {
      "type": 1,
      "components": [
        {
          "custom_id": "message:0:select:3",
          "type": 3,
          "options": [
            {
              "value": "1",
              "label": "1"
            },
            {
              "value": "2",
              "label": "2"
            },
            {
              "value": "3",
              "label": "3",
              "default": true
            }
          ]
        }
      ]
    }
  ],
  "flags": 64
}"
`;
